docs/internal/foreign_code.md

```diff
@@ -62,32 +65,210 @@ We hope that the region abstraction will make programming with sandboxes natural
+They have no `Self` parameter.
+All arguments are either `readonly` (if `const` in the C source) or `mut` otherwise.
+All types for C functions are types that are exposed for the specific sandbox object.
+
+C function pointers may be used to provide callbacks that invoke Verona code.
+
+### C globals
+
+C globals are defined as fields on the class that is exposed for the library.
+`const` C globals are surfaced as `imm` fields, all others as are `mut` in the region defined by an instance of the library.
+
+Mapping C++ constructs to Verona
+--------------------------------
+
+Two things are explicitly out of scope from Verona's C++ interop layer:
+
+ - Subclassing a C++ class in Verona.
+ - Subclassing a Verona class in C++.
+
+It *is* possible to invoke callbacks from C++ that invoke Verona code and so it is possible to create, on the C++ side of the interop layer, a C++ class that subclasses another C++ class and invokes Verona callbacks for all (or some) of its methods.
+Most of the rest of C++ should be possible to expose as Verona concepts.
+
+C++ namespaces
+--------------
+
+C++ namespaces are singleton nested classes within the class exposed for the library.
+Everything within the namespace is exposed within this context in the same way that it would be if it were declared in the top-level class.
+For example:
+
+```c++
+// In libfoo
+namespace SomeStuff
+{
+  int x;
+};
+```
+
+Imported into Verona looks like this:
 
 ```verona
-class LLVM
+import "libfoo" as F;
+
+var f = F();
+var x = f.SomeStuff.x : F.int;
+```
+
+### Classes
+
+C++ classes have a number of aspects that make them more complex than C structures.
+
+*Visibility specifiers* are respected by Verona code and there is no way to make a Verona class or function a `friend` of a C++ class.
+If you want to access a `private` field of a C++ object then this should be done by exposing a `friend` function on the C++ side of the interop layer.
+
+Inheritance is represented in Verona as each subclass defining a new concrete type.
+As with normal Verona code, there is *no* implicit conversion from a pointer to a C++ class to a pointer to any of its superclasses, C++ structures passed by pointer to C++ functions must be explicitly converted.
+This is one of the places where a programmer writing a wrapper library is likely to want to write some code to simplify the interface (this is not needed for correctness, only for the code to be user friendly).
+
+Verona's unified call syntax provides a mechanism for disambiguating fields in C++ classes that are in both a base class and a subclass.
+For example, consider this C++ code:
+
+```c++
+struct Super
 {
-	[#native, #interface_language="c++"]
-	imports library("libLLVM.so", "llvm.h");
-}
+  void x();
+};
+struct Sub : public Super
+{
+  void x();
+};
+```
+
+In Verona, given an instance of `Sub`, you could call `Super::x()` like this:
+
+```verona
+var s = CXXLib.Sub();
+CXXLib.Super.x(CXXLib.static_cast[CXXLib.Sub](s));
```

Comment:
This is clear, but not trivial. It involves users understanding both Verona and C++ concepts to a deeper level than I assume most people trying to interface the languages will have.

I'm guessing the majority of people using C++ inside Verona will be only well versed in one of the two languages. For example:
* A Verona programmer trying to speed up the program by using an optimised C++ library
* A C++ programmer trying to create a safer facade in Verona for a low-level C++ library

On the other hand, trying to sugar this will probably lead to many corner cases and impedance mismatch.

Reply:
Good point.  I'm not sure exactly who would work with this.  If it ends up being anything like the Python ecosystem, then there will be people writing Verona wrappers who will deal with this and so we'll end up with:

 - People who just want to grab some C/C++ code quickly working here with some quick-and-dirty code that won't really care much about the difficult corner cases because they're using fairly simple interfaces.
 - People writing idiomatic Verona wrappers around popular C/C++ libraries.  They'll hopefully be pretty knowledgeable about both Verona and C++ and we need to give them tools to write good bindings because they're going to end up writing things that a lot of people use.
 - People writing C/C++ specifically for use in the Verona context because they can write something faster in C/C++.  I hope there aren't many of these because in most cases I'd hope that the Verona compiler can do as good a job as the C/C++ one, but there may be some things where if you're willing to aggressively violate type safety then you can pool memory and make something faster in C/C++ than Verona.

The middle category are the folks I care the most about.  I want to make sure that they have to write the absolute minimum of boilerplate to be able to write idiomatic Verona wrappers around C/C++ libraries.  In an ideal world, they'll be able to quickly write a Verona wrapper that's nicer to use than the underlying library and so people writing new code would prefer to use libWhatever from Verona than from C/C++.